libmamba/src/core/satisfiability_error.cpp

```diff
@@ -0,0 +1,433 @@
+        {
+            // TODO source is recomputed.
+            // TODO: Once the new error message are not experimental, we should consider
+            // reducing this level since it is not somethig the user has control over.
+            LOG_WARNING << "Unexpected empty optionals for problem type "
+                        << solver_rule_name(problem.type)
+                        << "\nSource: " << pkg_info_to_str(problem.source())
+                        << "\nTarget: " << pkg_info_to_str(problem.target())
+                        << "\nDep: " << problem.dep().value_or("(empty)");
+        }
+
+        std::optional<ProblemsGraph::ProblemType> map_problem(SolverRuleinfo solverRuleInfo)
+        {
+            switch (solverRuleInfo)
+            {
+                case SOLVER_RULE_PKG_CONSTRAINS:
+                case SOLVER_RULE_PKG_REQUIRES:
+                case SOLVER_RULE_JOB:
+                case SOLVER_RULE_PKG:
+                case SOLVER_RULE_UPDATE:
+                    return std::nullopt;
+                case SOLVER_RULE_JOB_NOTHING_PROVIDES_DEP:
+                case SOLVER_RULE_PKG_NOTHING_PROVIDES_DEP:
+                case SOLVER_RULE_JOB_UNKNOWN_PACKAGE:
+                    return ProblemsGraph::ProblemType::NOT_FOUND;
+                case SOLVER_RULE_PKG_CONFLICTS:
+                case SOLVER_RULE_PKG_SAME_NAME:
+                    return ProblemsGraph::ProblemType::CONFLICT;
+                case SOLVER_RULE_BEST:
+                    return ProblemsGraph::ProblemType::BEST_NOT_INSTALLABLE;
+                case SOLVER_RULE_BLACK:
+                    return ProblemsGraph::ProblemType::ONLY_DIRECT_INSTALL;
+                case SOLVER_RULE_INFARCH:
+                    return ProblemsGraph::ProblemType::INFERIOR_ARCH;
+                case SOLVER_RULE_PKG_NOT_INSTALLABLE:
+                    return ProblemsGraph::ProblemType::NOT_INSTALLABLE;
+                case SOLVER_RULE_STRICT_REPO_PRIORITY:
+                    return ProblemsGraph::ProblemType::EXCLUDED_BY_REPO_PRIORITY;
+                case SOLVER_RULE_JOB_PROVIDED_BY_SYSTEM:
+                    return ProblemsGraph::ProblemType::PROVIDED_BY_SYSTEM;
+                case SOLVER_RULE_CHOICE:
+                case SOLVER_RULE_FEATURE:
+                case SOLVER_RULE_JOB_UNSUPPORTED:
+                case SOLVER_RULE_LEARNT:
+                case SOLVER_RULE_PKG_RECOMMENDS:
+                case SOLVER_RULE_RECOMMENDS:
+                case SOLVER_RULE_UNKNOWN:
+                    return std::nullopt;
+                case SOLVER_RULE_PKG_SELF_CONFLICT:
+                case SOLVER_RULE_PKG_OBSOLETES:
+                case SOLVER_RULE_PKG_IMPLICIT_OBSOLETES:
+                case SOLVER_RULE_PKG_INSTALLED_OBSOLETES:
+                case SOLVER_RULE_YUMOBS:
+                case SOLVER_RULE_DISTUPGRADE:
+                    return std::nullopt;
+            }
+            return std::nullopt;
+        }
+
+        class ProblemsGraphCreator
```

Comment:
don't we need to declare it in the header file in order to use it ?

Reply:
No, because we do not use it outside `satisfiability_error.cpp`. It's an implementation detail only used inside `ProblemsGraph::from_solver` but nowhere else.